File: edge-server/src/db/models/order.rs
//! Order Model (Graph Model)
//!
//! 归档订单使用图边关系，只存储核心数据。
//! 所有订单变更通过 OrderManager 事件溯源处理。

use super::serde_helpers;
use serde::{Deserialize, Serialize};
use surrealdb::RecordId;

// =============================================================================
// Archived Order Types (used by archive service)
// =============================================================================

/// Order status enum (archived orders only)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum OrderStatus {
    Completed,
    Void,
    Moved,
    Merged,
}

/// Archived order entity
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Order {
    #[serde(default, with = "serde_helpers::option_record_id")]
    pub id: Option<RecordId>,
    pub receipt_number: String,
    pub zone_name: Option<String>,
    pub table_name: Option<String>,
    pub status: OrderStatus,
    pub is_retail: bool,
    pub guest_count: Option<i32>,
    pub original_total: f64,
    pub subtotal: f64,
    pub total_amount: f64,
    pub paid_amount: f64,
    pub discount_amount: f64,
    pub surcharge_amount: f64,
    pub comp_total_amount: f64,
    pub order_manual_discount_amount: f64,
    pub order_manual_surcharge_amount: f64,
    pub tax: f64,
    pub start_time: i64,
    pub end_time: Option<i64>,
    pub operator_id: Option<String>,
    pub operator_name: Option<String>,
    #[serde(default, with = "serde_helpers::option_record_id")]
    pub related_order_id: Option<RecordId>,
    // === Void Metadata (only when status == Void) ===
    /// 作废类型: "CANCELLED" | "LOSS_SETTLED"
    #[serde(skip_serializing_if = "Option::is_none")]
    pub void_type: Option<String>,
    /// 损失原因: "CUSTOMER_FLED" | "CUSTOMER_INSOLVENT" | "OTHER"
    #[serde(skip_serializing_if = "Option::is_none")]
    pub loss_reason: Option<String>,
    /// 损失金额（未收回部分）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub loss_amount: Option<f64>,
    /// 作废备注
    #[serde(skip_serializing_if = "Option::is_none")]
    pub void_note: Option<String>,
    pub prev_hash: String,
    pub curr_hash: String,
    pub created_at: Option<i64>,
}

/// Split item in a payment
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct SplitItem {
    pub instance_id: String,
    pub name: String,
    pub quantity: i32,
    #[serde(default)]
    pub unit_price: f64,
}

// =============================================================================
// API Response Types (for frontend)
// =============================================================================

fn default_guest_count() -> i32 {
    1
}

/// Order summary for list view
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrderSummary {
    pub order_id: String,
    pub receipt_number: String,
    pub table_name: Option<String>,
    pub status: String,
    #[serde(default)]
    pub is_retail: bool,
    #[serde(default)]
    pub total: f64,
    #[serde(default = "default_guest_count")]
    pub guest_count: i32,
    #[serde(default)]
    pub start_time: i64,
    pub end_time: Option<i64>,
    // === Void Metadata ===
    #[serde(default)]
    pub void_type: Option<String>,
    #[serde(default)]
    pub loss_reason: Option<String>,
    #[serde(default)]
    pub loss_amount: Option<f64>,
}

/// Order item option for detail view
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrderItemOptionDetail {
    pub attribute_name: String,
    pub option_name: String,
    pub price_modifier: f64,
}

/// Order item for detail view
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrderItemDetail {
    pub id: String,
    pub instance_id: String,
    pub name: String,
    pub spec_name: Option<String>,
    pub category_name: Option<String>,
    #[serde(default)]
    pub price: f64,
    #[serde(default)]
    pub quantity: i32,
    #[serde(default)]
    pub unpaid_quantity: i32,
    #[serde(default)]
    pub unit_price: f64,
    #[serde(default)]
    pub line_total: f64,
    #[serde(default)]
    pub discount_amount: f64,
    #[serde(default)]
    pub surcharge_amount: f64,
    pub note: Option<String>,
    #[serde(default)]
    pub selected_options: Vec<OrderItemOptionDetail>,
}

/// Payment for detail view
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrderPaymentDetail {
    #[serde(default)]
    pub payment_id: Option<String>,
    pub method: String,
    pub amount: f64,
    pub timestamp: i64,
    #[serde(default)]
    pub cancelled: bool,
    pub cancel_reason: Option<String>,
    #[serde(default)]
    pub split_type: Option<String>,
    #[serde(default, deserialize_with = "deserialize_split_items")]
    pub split_items: Vec<SplitItem>,
    #[serde(default)]
    pub aa_shares: Option<i32>,
    #[serde(default)]
    pub aa_total_shares: Option<i32>,
}

/// Deserialize JSON string to Vec<SplitItem>
fn deserialize_split_items<'de, D>(deserializer: D) -> Result<Vec<SplitItem>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let opt: Option<String> = Option::deserialize(deserializer)?;
    match opt {
        Some(s) if !s.is_empty() => {
            serde_json::from_str(&s).map_err(serde::de::Error::custom)
        }
        _ => Ok(Vec::new()),
    }
}

/// Event for detail view
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrderEventDetail {
    pub event_id: String,
    pub event_type: String,
    pub timestamp: i64,
    #[serde(deserialize_with = "deserialize_json_string")]
    pub payload: Option<serde_json::Value>,
}

/// Deserialize JSON string to Value
fn deserialize_json_string<'de, D>(deserializer: D) -> Result<Option<serde_json::Value>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let opt: Option<String> = Option::deserialize(deserializer)?;
    match opt {
        Some(s) if !s.is_empty() => {
            serde_json::from_str(&s).map(Some).map_err(serde::de::Error::custom)
        }
        _ => Ok(Some(serde_json::Value::Object(serde_json::Map::new()))),
    }
}

/// Full order detail (for frontend)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrderDetail {
    pub order_id: String,
    pub receipt_number: String,
    pub table_name: Option<String>,
    pub zone_name: Option<String>,
    pub status: String,
    #[serde(default)]
    pub is_retail: bool,
    #[serde(default = "default_guest_count")]
    pub guest_count: i32,
    #[serde(default)]
    pub total: f64,
    #[serde(default)]
    pub paid_amount: f64,
    #[serde(default)]
    pub total_discount: f64,
    #[serde(default)]
    pub total_surcharge: f64,
    #[serde(default)]
    pub comp_total_amount: f64,
    #[serde(default)]
    pub order_manual_discount_amount: f64,
    #[serde(default)]
    pub order_manual_surcharge_amount: f64,
    #[serde(default)]
    pub start_time: i64,
    pub end_time: Option<i64>,
    pub operator_name: Option<String>,
    // === Void Metadata ===
    #[serde(default)]
    pub void_type: Option<String>,
    #[serde(default)]
    pub loss_reason: Option<String>,
    #[serde(default)]
    pub loss_amount: Option<f64>,
    #[serde(default)]
    pub void_note: Option<String>,
    #[serde(default)]
    pub items: Vec<OrderItemDetail>,
    #[serde(default)]
    pub payments: Vec<OrderPaymentDetail>,
    #[serde(default)]
    pub timeline: Vec<OrderEventDetail>,
}

File: crab-console/src/core/types/order.ts
export interface OrderSummary {
  id: number;
  source_id: string;
  receipt_number: string | null;
  status: string;
  end_time: number | null;
  total: number | null;
  synced_at: number;
}

export type ChainEntryType = 'ORDER' | 'CREDIT_NOTE' | 'ANULACION' | 'UPGRADE';

export interface ChainEntryItem {
  entry_type: ChainEntryType;
  entry_id: number;
  display_number: string;
  status: string;
  amount: number | null;
  created_at: number;
  original_order_id: number | null;
  original_receipt: string | null;
}

export interface AnulacionDetailResponse {
  order_id: number;
  receipt_number: string;
  total_amount: number;
  is_anulada: boolean;
  created_at: number;
}

export interface UpgradeDetailResponse {
  order_id: number;
  receipt_number: string;
  total_amount: number;
  tax: number;
  is_upgraded: boolean;
  customer_nif: string | null;
  customer_nombre: string | null;
  customer_address: string | null;
  customer_email: string | null;
  customer_phone: string | null;
  created_at: number;
}

export interface CreditNoteDetailResponse {
  source_id: number;
  credit_note_number: string;
  original_order_id: number;
  original_receipt: string;
  subtotal_credit: number;
  tax_credit: number;
  total_credit: number;
  refund_method: string;
  reason: string;
  note: string | null;
  operator_name: string;
  authorizer_name: string | null;
  created_at: number;
  items: CreditNoteItem[];
}

export interface CreditNoteItem {
  item_name: string;
  quantity: number;
  unit_price: number;
  line_credit: number;
  tax_rate: number;
  tax_credit: number;
}

export interface OrderItemOption {
  attribute_name: string;
  option_name: string;
  price: number;
  quantity: number;
}

export interface OrderItem {
  name: string;
  spec_name: string | null;
  category_name: string | null;
  price: number;
  quantity: number;
  unit_price: number;
  line_total: number;
  discount_amount: number;
  surcharge_amount: number;
  tax: number;
  tax_rate: number;
  is_comped: boolean;
  note: string | null;
  options: OrderItemOption[];
}

export interface OrderPayment {
  seq: number;
  method: string;
  amount: number;
  timestamp: number;
  cancelled: boolean;
}

export interface TaxDesglose {
  tax_rate: number;
  base_amount: number;
  tax_amount: number;
}

export interface OrderEvent {
  seq: number;
  event_type: string;
  timestamp: number;
  operator_id: number | null;
  operator_name: string | null;
  data: string | null;
}

export interface OrderDetailPayload {
  zone_name: string | null;
  table_name: string | null;
  is_retail: boolean;
  guest_count: number | null;
  original_total: number;
  subtotal: number;
  paid_amount: number;
  discount_amount: number;
  surcharge_amount: number;
  comp_total_amount: number;
  order_manual_discount_amount: number;
  order_manual_surcharge_amount: number;
  order_rule_discount_amount: number;
  order_rule_surcharge_amount: number;
  start_time: number;
  operator_name: string | null;
  void_type: string | null;
  loss_reason: string | null;
  loss_amount: number | null;
  void_note: string | null;
  member_name: string | null;
  items: OrderItem[];
  payments: OrderPayment[];
  events?: OrderEvent[];
}

export interface OrderDetailResponse {
  source: string;
  detail: OrderDetailPayload;
  desglose: TaxDesglose[];
}

export interface CreditNoteSummary {
  source_id: number;
  credit_note_number: string;
  total_credit: number;
  refund_method: string;
  reason: string;
  operator_name: string;
  created_at: number;
}
